src/common/graph/sieve.js

```diff
@@ -93,34 +95,38 @@ function updateNode(current, index, change, result) {
   if (isBranch(change) && isBranch(node)) {
     // Both are branches: Recursively merge children.
     const nextResult = [];
+    node.version = change.version;
     sieve(node.children, change.children, nextResult);
     if (nextResult.length) result.push({ ...change, children: nextResult });
   } else if (isBranch(node)) {
     // Current node is a branch but the change is a leaf; if the branch
     // has newer children, ignore the change and keep only those children;
     // Otherwise, discard the branch and keep the change.
-    const newNode = getNewer(node, change.version);
+    const newNode = getNewer(node, change);
     current[index] = newNode || change;
     if (!newNode) result.push(change);
     // TODO: In the case of partial removal, what should result be?
   } else {
     // Current node is a leaf. Replace with the change if it is newer.
-    const newChange = getNewer(change, node.version);
+    const newChange = getNewer(change, node);
     if (newChange) {
       current[index] = newChange;
-      result.push(newChange);
+      // console.log(current);
+      if (change.value !== node.value || change.path !== node.path) {
+        result.push(newChange);
+      }
```

Comment:
This is the main piece of this PR. Don't push a leaf node into the result if nothing changed.

Branches in the result won't be created if there are no leaves.